File: 10-28-20/fib.cpp
// To run:
// - say a little prayer
// - compile with: g++ -std=c++11 fib.cpp -o fib.exe
// - run with: ./fib.exe

#include <iostream>
#include <map>

using namespace std;

int fib(int n, map<int, int>* memo) {
  if (memo->count(n) == 1)
    return memo->at(n);

  if (n == 1 || n == 2)
    return 1;

  int result = fib(n - 1, memo) + fib(n - 2, memo);
  memo->insert(pair<int, int>(n, result));
  return result;
}

int fib(int n) {
  map<int, int> newMemo {};
  return fib(n, &newMemo);
}

int main() {
  cout << fib(6) << endl; // 8
  cout << fib(7) << endl; // 13
  cout << fib(8) << endl; // 21
  cout << fib(45) << endl; // 1134903170
}

File: 10-28-20/fib.rb
# def fib(n)
#   return 1 if n == 1 || n == 2
#   return fib(n - 1) + fib(n - 2)
# end

def fib(n, memo={})
  return memo[n] if memo.has_key?(n)
  return 1 if n == 1 || n == 2
  memo[n] = fib(n - 1, memo) + fib(n - 2, memo)
  return memo[n]
end

p fib(6) # 8
p fib(7) # 13
p fib(8) # 21
p fib(45) # 1134903170
